oauth2_http/java/com/google/auth/oauth2/JwtCredentials.java

```diff
@@ -0,0 +1,334 @@
+  }
+
+  /**
+   * Returns a copy of these credentials with modified claims.
+   *
+   * @param newClaims New claims. Any unspecified claim fields will default to the the current
+   *        values.
+   * @return new credentials
+   */
+  @Override
+  public JwtCredentials jwtWithClaims(Claims newClaims) {
+    return JwtCredentials.newBuilder()
+        .setPrivateKey(privateKey)
+        .setPrivateKeyId(privateKeyId)
+        .setClaims(claims.merge(newClaims))
+        .build();
+  }
+
+  @Override
+  public String getAuthenticationType() {
+    return "JWT";
+  }
+
+  @Override
+  public Map<String, List<String>> getRequestMetadata(URI uri) throws IOException {
+    synchronized (lock) {
+      if (shouldRefresh()) {
+        refresh();
+      }
+      List<String> newAuthorizationHeaders = Collections.singletonList(JWT_ACCESS_PREFIX + jwt);
+      return Collections.singletonMap(AuthHttpConstants.AUTHORIZATION, newAuthorizationHeaders);
+    }
+  }
+
+  @Override
+  public boolean hasRequestMetadata() {
+    return true;
+  }
+
+  @Override
+  public boolean hasRequestMetadataOnly() {
+    return true;
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (!(obj instanceof JwtCredentials)) {
+      return false;
+    }
+    JwtCredentials other = (JwtCredentials) obj;
+    return Objects.equals(this.privateKey, other.privateKey)
+        && Objects.equals(this.privateKeyId, other.privateKeyId)
+        && Objects.equals(this.claims, other.claims)
+        && Objects.equals(this.lifeSpanSeconds, other.lifeSpanSeconds);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(this.privateKey, this.privateKeyId, this.claims, this.lifeSpanSeconds);
+  }
+
+  Clock getClock() {
+    if (clock == null) {
+      clock = Clock.SYSTEM;
+    }
+    return clock;
+  }
+
+  public static class Builder {
+    private PrivateKey privateKey;
+    private String privateKeyId;
+    private Claims claims;
+    private Clock clock = Clock.SYSTEM;
+    private Long lifeSpanSeconds = TimeUnit.HOURS.toSeconds(1);
+
+    protected Builder() {}
+
+    public Builder setPrivateKey(PrivateKey privateKey) {
+      this.privateKey = privateKey;
+      return this;
+    }
+
+    public PrivateKey getPrivateKey() {
+      return privateKey;
+    }
+
+    public Builder setPrivateKeyId(String privateKeyId) {
+      this.privateKeyId = privateKeyId;
+      return this;
+    }
+
+    public String getPrivateKeyId() {
+      return privateKeyId;
+    }
+
+    public Builder setClaims(Claims claims) {
+      this.claims = claims;
+      return this;
+    }
+
+    public Claims getClaims() {
+      return claims;
+    }
+
+    public Builder setLifeSpanSeconds(Long lifeSpanSeconds) {
+      this.lifeSpanSeconds = lifeSpanSeconds;
+      return this;
+    }
+
+    public Long getLifeSpanSeconds() {
+      return lifeSpanSeconds;
+    }
+
+    Builder setClock(Clock clock) {
+      this.clock = clock;
```

Comment:
Is there a reason you allow null clocks and set it to system in the getter? Doing it here feels more thread safe.

Reply:
We shouldn't be allowing null clocks. I will add some `Preconditions` checks to fail early.

The null checking in `JwtCredentials#getClock()` is to satisfy the `Serializable` interface - clocks are not serializable. When we deserialize a `JwtCredentials` instance, it's going to default back to the system clock.